File: typescript-nodejs/architecture-shell-cqrs/src/ICommandHandler.ts
import { Command } from './Command';
import { IRequestHandler } from './IRequestHandler';
import { BaseRequest } from './BaseRequest';

/**
 * ICommandHandler - Processes commands without return values
 *
 * Handles state-changing operations that return void (Unit equivalent).
 *
 * @typeParam TCommand - The command type (must extend Command)
 *
 * @remarks
 * - May return Result<void> if business validation errors expected
 * - Infrastructure errors always throw exceptions
 * - Executes within transaction boundary (managed by UnitOfWork behavior)
 *
 * @example
 * ```typescript
 * export class PublishBlogPostHandler implements ICommandHandler<PublishBlogPostCommand> {
 *   async handle(
 *     command: PublishBlogPostCommand,
 *     signal: AbortSignal
 *   ): Promise<void> {
 *     // Publish blog post logic
 *   }
 * }
 * ```
 *
 * @see {@link IQueryHandler} - For read-only operations
 */
export interface ICommandHandler<TCommand extends Command>
  extends IRequestHandler<TCommand, void> {}

/**
 * ICommandHandlerWithResult - Processes commands with return values
 *
 * Handles state-changing operations that return data
 * (e.g., created entity ID, updated version number).
 *
 * @typeParam TCommand - The command type (extends BaseRequest)
 * @typeParam TResult - The return type (typically Result<TValue>)
 *
 * @example
 * ```typescript
 * export class CreateOrderHandler implements ICommandHandlerWithResult<CreateOrderCommand, Result<string>> {
 *   async handle(
 *     command: CreateOrderCommand,
 *     signal: AbortSignal
 *   ): Promise<Result<string>> {
 *     // Create order logic
 *     return Result.success(order.id);
 *   }
 * }
 * ```
 */
export interface ICommandHandlerWithResult<TCommand extends BaseRequest<TResult>, TResult>
  extends IRequestHandler<TCommand, TResult> {}

File: csharp-dotnet/src/Architecture.Shell.Cqrs/ICommandHandler.cs
using Architecture.Core.Functional;

namespace Architecture.Shell.Cqrs;

/// <summary>
/// Handler for commands that modify state without return values (void commands).
/// </summary>
/// <typeparam name="TCommand">The command type.</typeparam>
/// <remarks>
/// Returns <see cref="Result"/> to support business validation errors.
/// Use <see cref="Result.Success()"/> for successful completion or
/// <see cref="Result.Failure"/> for business rule violations.
/// </remarks>
public interface ICommandHandler<in TCommand> : IRequestHandler<TCommand, Result>
    where TCommand : ICommand
{
}

/// <summary>
/// Handler for commands that modify state and return a value.
/// </summary>
/// <typeparam name="TCommand">The command type.</typeparam>
/// <typeparam name="TResult">
/// The return type, commonly <see cref="Result{T}"/> for mixed error handling.
/// </typeparam>
public interface ICommandHandler<in TCommand, TResult> : IRequestHandler<TCommand, TResult>
    where TCommand : ICommand<TResult>
{
}
